### PR TITLE
[C#] overflow new exercise

### PR DESCRIPTION
**Pull request by mikedamay**

### CODE REVIEW

**Comment by mikedamay** [comment]
Please review _Example.cs_, _instructions.md_ and _introduction.md_.

All other files except _csharp/config.json_ are complete apart from proofing and show the direction of travel.

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
+The country of Hyperia has its problems. In particular, inflation is out of control. You have been flown in to take up a contract at the central bank.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:1]: I like this! :D

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+| Base    | Multiplier | Denomination |
+| ------- | ---------- | ------------ |
+| 10,000  | 1,000      | 10,000,000   |
+| 50,000  | 1,000      | 50,000,000   |
+| 100,000 | 1,000      | 100,000,000  |
+| ...     | 1,000      | ...          |
+
+In summary, this means that _denomination_ = _base_ \* _multiplier_.
+
+Please implement the method `CentrlBank.DisplayDenomination` which takes the base and multiplier and returns a string of their product (base \* multiplier). The numeric string returned should not be formatted.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:20]:
```suggestion
Please implement the method `CentralBank.DisplayDenomination()` which takes the base and multiplier and returns a string of their product (base \* multiplier). The numeric string returned should not be formatted.
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+If the multiplication operation causes an overflow then the string, "\*\*\* Too Big \*\*\*", should be returned.
+
+```csharp
+CentralBank.DisplayDenomination(10000L, 1000L);
+// => 10000000
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:26]:
```suggestion
// => "10000000"
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+The central bank uses the same multiplier approach to recording GDP internally. Statistics are recorded in a base denomination and then scaled up using the multiplier.
+
+Please implement the method `CentrlBank.DisplayGDP` which takes the base GDP and the current multiplier and returns the GDP as a string.
+
+If the GDP cannot be calculated then "\*\*\* Too Big \*\*\*" is returned.
+
+```csharp
+CentralBank.DisplayGDP(555f, 10000f);
+// => 555000000
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:41]:
```suggestion
// => "555000000"
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+| 10,000  | 1,000      | 10,000,000   |
+| 50,000  | 1,000      | 50,000,000   |
+| 100,000 | 1,000      | 100,000,000  |
+| ...     | 1,000      | ...          |
+
+In summary, this means that _denomination_ = _base_ \* _multiplier_.
+
+Please implement the method `CentrlBank.DisplayDenomination` which takes the base and multiplier and returns a string of their product (base \* multiplier). The numeric string returned should not be formatted.
+
+If the multiplication operation causes an overflow then the string, "\*\*\* Too Big \*\*\*", should be returned.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:22]: Maybe highlighting the string as code to make it stand out a bit more?

```suggestion
If the multiplication operation causes an overflow then the string, `"*** Too Big ***"`, should be returned.
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+// => "*** Too Big ***"
+```
+
+## 2. Display GDP for Hyperia
+
+The central bank uses the same multiplier approach to recording GDP internally. Statistics are recorded in a base denomination and then scaled up using the multiplier.
+
+Please implement the method `CentrlBank.DisplayGDP` which takes the base GDP and the current multiplier and returns the GDP as a string.
+
+If the GDP cannot be calculated then "\*\*\* Too Big \*\*\*" is returned.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:37]: Maybe highlighting the string as code to make it stand out a bit more?

```suggestion
If the GDP cannot be calculated then `"*** Too Big ***"` is returned.
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+// => 10000000
+CentralBank.DisplayDenomination(long.MaxValue / 2, 10000L);
+// => "*** Too Big ***"
+```
+
+## 2. Display GDP for Hyperia
+
+The central bank uses the same multiplier approach to recording GDP internally. Statistics are recorded in a base denomination and then scaled up using the multiplier.
+
+Please implement the method `CentrlBank.DisplayGDP` which takes the base GDP and the current multiplier and returns the GDP as a string.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:35]:
```suggestion
Please implement the method `CentralBank.DisplayGDP()` which takes the base GDP and the current multiplier and returns the GDP as a string.
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+// => "*** Too Big ***"
+```
+
+## 3. Display Chief Economist's salary
+
+Your boss at the bank wants to make sure there will be no problems at payroll time.
+
+Implement the `CentralBank.DisplayChiefEconomistSalary()` method that takes a salary expressed in base units (as above) and a multiplier and returns a string containing the product.
+
+If the salary cannot be calculated then "\*\*\* Much Too Big \*\*\*" is returned.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:52]:
```suggestion
If the salary cannot be calculated then `"*** Much Too Big ***"` is returned.
```

File: languages/csharp/exercises/concept/overflow/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+```csharp
+CentralBank.DisplayChiefEconomistSalary(555000m, 10000m);
+// => "5550000000"
+CentralBank.DisplayChiefEconomistSalary(555000m,decimal.MaxValue / 2L);
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/instructions.md:57]:
```suggestion
CentralBank.DisplayChiefEconomistSalary(555000m, decimal.MaxValue / 2L);
```

**Comment by mikedamay** [comment]
closed to finalize - will reopen shortly

**Comment by mikedamay** [comment]
All files complete and proofed except _csharp/config.json_

File: languages/csharp/exercises/concept/overflow/.docs/after.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+```csharp
+var xx = new Random().Next();
+checked
+{
+    int expr = int.MaxValue * xx;  // overflow exception is thrown
+}
+
+// or
+
+int expr2 = checked(int.MaxValue * xx);  // overflow exception is thrown
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/after.md:14]: We could get rid of the `Random` bit by introducing a variable, with which we can increment by one to hopefully even more clearly show the overflow:

```suggestion
int one = 1;
checked
{
    int expr = int.MaxValue + one;  // overflow exception is thrown
}

// or

int expr2 = checked(int.MaxValue + one);  // overflow exception is thrown
```

File: languages/csharp/exercises/concept/overflow/.docs/after.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
+The exercise shows the behavior of various numeric types when they overflow, i.e. when their capacity is insufficient to contain the value resulting from a computation such as an arithmetic operation or cast.
+
+- unsigned [integers][integral-numeric-types] (`byte`, `ushort`, `uint`, `ulong`) will wrap around to zero (the type's maximum value + 1 acts as a modulus) unless [broadly speaking][checked-compiler-setting] they appear within a [`checked`][checked-and-unchecked] block in which case an instance of `OverflowException` is thrown. `int` and `long` will behave similarly except that they wrap around to `int.MinValue` and `long.minValue` respectively.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/after.md:3]: When testing things, I just found at that the C# compiler can detect overflows when using constants:

```csharp
var expr = int.MaxValue + 1;
```

This code leads to a compiler error:

>  Program.cs(19, 28): [CS0220] The operation overflows at compile time in checked mode

Might be nice to mention this somewhere too.

File: languages/csharp/exercises/concept/overflow/.docs/after.md [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    int expr = int.MaxValue * xx;  // overflow exception is thrown
+}
+
+// or
+
+int expr2 = checked(int.MaxValue * xx);  // overflow exception is thrown
+```
+
+- the `checked` state applies only to expressions directly in the block. Overflow states in called functions are not caught.
+- [`float` and `double`][floating-point-numeric-types] types will adopt a state of _infinity_ that can be tested wtih `float.IsInfinity()` etc.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/after.md:18]: Maybe we can add a floating overflow example too?

File: languages/csharp/exercises/concept/overflow/.docs/after.md [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+- Numbers of type [`decimal`][floating-point-numeric-types] will cause an instance of `OverflowException` to be thrown.
+- There is a corresponding `unchecked` keyword for circumstances where you want to reverse the effect of `checed` inside a `checked` block or when the compiler setting has been used.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/after.md:20]:
```suggestion
- There is a corresponding `unchecked` keyword for circumstances where you want to reverse the effect of `unchecked` inside a `checked` block or when the compiler setting has been used.
```

Comment (ErikSchierboom) [reply languages/csharp/exercises/concept/overflow/.docs/after.md:20]: Maybe have an example for this?

File: languages/csharp/exercises/concept/overflow/.docs/after.md [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+Overflows that occur without an exception being thrown can be problematic because it's generally true that the earlier an error condition can be reported the better.
+
+Problems with overflows for `int` and `float` can be mitigated by assigning the result to a variable of type `long`, `decimal` or `double`.
+
+If large integers are essential to your code then using the [`BigInteger`][big-integer] type is an option.
+
+Naturally there are occasions on which it is legitimate to allow an integer to wrap around particularly in the case of unsigned values. A classic case is that of hash codes that use the width of the integer as a kind of modulo.
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/after.md:28]: Really nice example!

File: languages/csharp/exercises/concept/overflow/Overflow.cs [hunk 15]
```diff
@@ -0,0 +1,19 @@
+using System;
+
+public static class CentralBank
+{
+    public static string DisplayDenomination(long @base, long multiplier)
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/Overflow.cs:5]: Note: we have to explain the `@` prefix for identifiers somewhere in a prerequisite exercise. Might be something to add to the `after.md` of the `basics` exercise.

File: languages/csharp/exercises/concept/overflow/.docs/introduction.md [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+```csharp
+var xx = new Random().Next();
+checked
+{
+    int expr = int.MaxValue * xx;   // OverflowException is thrown
+}
+
+// or
+
+int expr2 = checked(int.MaxValue * xx);     // OverflowException is thrown
```
Comment (ErikSchierboom) [inline languages/csharp/exercises/concept/overflow/.docs/introduction.md:16]: We could get rid of the `Random` bit by introducing a variable, with which we can increment by one to hopefully even more clearly show the overflow:

```suggestion
int one = 1;
checked
{
    int expr = int.MaxValue + one;   // OverflowException is thrown
}

// or

int expr2 = checked(int.MaxValue + one);     // OverflowException is thrown
```

File: languages/csharp/exercises/concept/overflow/Overflow.cs (hunk 15 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/overflow/Overflow.cs:5]: Issue #1979 addresses this.

File: languages/csharp/exercises/concept/overflow/.docs/after.md (hunk 12 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/overflow/.docs/after.md:19]: Done

File: languages/csharp/exercises/concept/overflow/.docs/after.md (hunk 11 above)
Comment (mikedamay) [reply languages/csharp/exercises/concept/overflow/.docs/after.md:3]: Done

**Pull request merged**